src/infra/core/src/repos/dataset_repository_s3.rs

```diff
@@ -123,94 +150,119 @@ impl DatasetRepository for DatasetRepositoryS3 {
                 while let Some(hdl) = datasets.next().await {
                     let hdl = hdl?;
-                    if hdl.id == *id {
+                    if hdl.alias == normalized_alias {
                         return Ok(hdl);
                     }
                 }
                 Err(GetDatasetError::NotFound(DatasetNotFoundError {
                     dataset_ref: dataset_ref.clone(),
                 }))
             }
+            DatasetRef::ID(id) => {
+                if self
+                    .s3_context
+                    .bucket_path_exists(&id.cid.to_string())
+                    .await?
+                {
+                    let dataset = self.get_dataset_impl(id).await?;
+                    let dataset_alias = self.resolve_dataset_alias(&dataset).await?;
+                    Ok(DatasetHandle::new(id.clone(), dataset_alias))
+                } else {
+                    Err(GetDatasetError::NotFound(DatasetNotFoundError {
+                        dataset_ref: dataset_ref.clone(),
+                    }))
+                }
+            }
         }
     }
 
     fn get_all_datasets<'s>(&'s self) -> DatasetHandleStream<'s> {
-        Box::pin(async_stream::try_stream! {
-            let folders_common_prefixes = self.s3_context.bucket_list_folders().await?;
-            for prefix in folders_common_prefixes {
-                let mut prefix = prefix.prefix.unwrap();
-                while prefix.ends_with('/') {
-                    prefix.pop();
-                }
-
-                let name = DatasetAlias::new(None, DatasetName::try_from(prefix).int_err()?);
-                let hdl = self.resolve_dataset_ref(&name.into()).await.int_err()?;
-                yield hdl;
-            }
-        })
+        self.stream_datasets_if(|_| true)
     }
 
     fn get_account_datasets<'s>(&'s self, account_name: AccountName) -> DatasetHandleStream<'s> {
-        if account_name == self.current_account_subject.account_name {
-            self.get_all_datasets()
-        } else {
+        if !self.is_multi_tenant() && account_name != self.current_account_subject.account_name {
             panic!("Single-tenant dataset repository queried by non-default account");
         }
+
+        self.stream_datasets_if(move |dataset_alias| {
+            if let Some(dataset_account_name) = &dataset_alias.account_name {
+                dataset_account_name == &account_name
+            } else {
+                true
+            }
+        })
     }
 
     async fn get_dataset(
         &self,
         dataset_ref: &DatasetRef,
     ) -> Result<Arc<dyn Dataset>, GetDatasetError> {
-        let handle = self.resolve_dataset_ref(dataset_ref).await?;
-        let dataset = self.get_dataset_impl(&handle.alias).await?;
+        let dataset_handle = self.resolve_dataset_ref(dataset_ref).await?;
+        let dataset = self.get_dataset_impl(&dataset_handle.id).await?;
         Ok(Arc::new(dataset))
     }
 
+    // TODO: name collision check
```

Comment:
It seems to be checking for name collisions below. TODO either lacking some details or can be removed.

Reply:
Killed. No longer relevant.